runtime/src/main/jni/ObjectManager.cpp

```diff
@@ -263,6 +286,56 @@ void ObjectManager::JSObjectWeakCallbackStatic(const WeakCallbackInfo<ObjectWeak
     thisPtr->JSObjectWeakCallback(isolate, callbackState);
 }
 
+void ObjectManager::JSObjectFinalizerStatic(const WeakCallbackInfo<ObjectWeakCallbackState>& data) {
+    ObjectWeakCallbackState* callbackState = data.GetParameter();
+
+    ObjectManager* thisPtr = callbackState->thisPtr;
+
+    auto isolate = data.GetIsolate();
+
+    thisPtr->JSObjectFinalizer(isolate, callbackState);
+}
+
+void ObjectManager::JSObjectFinalizer(Isolate* isolate, ObjectWeakCallbackState* callbackState) {
+    Persistent<Object>* po = callbackState->target;
+
+    auto jsInfoIdx = static_cast<int>(MetadataNodeKeys::JsInfo);
+    auto jsInstance = po->Get(m_isolate);
+    auto jsInfo = jsInstance->GetInternalField(jsInfoIdx);
+    if (jsInfo->IsUndefined()) {
+        // Typescript object layout has an object instance as child of the actual registered instance. checking for that
+        auto prototypeObject = jsInstance->GetPrototype().As<Object>();
+        if (!prototypeObject.IsEmpty() && prototypeObject->IsObject()) {
+            DEBUG_WRITE("GetJSInstanceInfo: need to check prototype :%d", prototypeObject->GetIdentityHash());
+            if (IsJsRuntimeObject(prototypeObject)) {
+                jsInfo = prototypeObject->GetInternalField(jsInfoIdx);
+            }
+        }
+    }
+
+    if (jsInfo.IsEmpty() || !jsInfo->IsExternal()) {
```

Comment:
ok in that case leave `IsEmpty` and `IsUndefined`, but in both cases they implicitly imply `!IsExternal`